pandas/core/indexing.py

```diff
@@ -669,6 +670,56 @@ def _get_setitem_indexer(self, key):
 
         return self._convert_to_indexer(key, axis=0)
 
+    @final
+    def _maybe_mask_setitem_value(self, indexer, value):
+        """
+        If we have obj.iloc[mask] = arraylike and arraylike has the same
+        length as obj, we treat this as obj.iloc[mask] = arraylike[mask],
+        similar to Series.__setitem__.
+        """
+        # TODO: why do we only do this for loc, not iloc?
+        # ser = pd.Series(range(5))
+        # mask = np.array([True, False, True, False, True])
+        # ser[mask] = ser * 2  # <- works
+        # ser.loc[mask] = ser * 2  # <- works
+        # ser.iloc[mask] = ser * 2  # <- raises
+
+        if isinstance(indexer, tuple) and len(indexer) == 2:
+            pi, icols = indexer
+            ndim = getattr(value, "ndim", 0)
+            if ndim == 0:
+                pass
+            elif com.is_bool_indexer(pi) and len(value) == len(pi):
+                newkey = pi.nonzero()[0]
+
+                if is_scalar_indexer(icols, self.ndim - 1) and ndim == 1:
+                    # e.g. test_loc_setitem_boolean_mask_allfalse
+                    value = value[pi]
+                    indexer = (newkey, icols)
+
+                elif (
+                    isinstance(icols, np.ndarray)
+                    and icols.dtype.kind == "i"
+                    and len(icols) == 1
+                ):
+                    if ndim == 1:
+                        value = value[pi]
+                        indexer = (newkey, icols)
+                        # TODO: is it sketchy that icols is an ndarray and value
```

Comment:
I think this is ok, something like

```
df = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})


df.loc[[False, True, True], ["a"]] = pd.Series([10, 20, 30])
```

is valid and runs through there

Reply:
i guess.  still seems weird as the analogous behavior with ndarray doesn't work

```
arr = np.random.randn(3, 2)
values = np.array([1, 2, 3])

arr[:, 0] = values  # <- works
arr[:, [0]] = values  # <- raises
```

Reply:
This is odd yes